File: python-examples/bs4-example/api/details.py
from playwright.async_api import Page, BrowserContext
from typing import TypedDict, List
from bs4 import BeautifulSoup
from intuned_browser import go_to_url


class Params(TypedDict):
    title: str
    price: str
    details_url: str


class ProductDetails(TypedDict):
    title: str
    price: str
    source_url: str
    description: str
    sku: str
    category: str
    sizes: List[str]
    colors: List[str]
    images: List[str]


def extract_product_details(html: str, params: Params) -> ProductDetails:
    """
    Extracts detailed product information from the page HTML using BeautifulSoup.
    Replace selectors with appropriate ones for your target site.
    """
    soup = BeautifulSoup(html, "html.parser")

    # Extract description - replace selector as needed
    description_element = soup.select_one("#tab-description p")
    description = (
        description_element.get_text(strip=True) if description_element else ""
    )

    # Extract SKU - replace selector as needed
    sku_element = soup.select_one(".sku")
    sku = sku_element.get_text(strip=True) if sku_element else ""

    # Extract category - replace selector as needed
    category_element = soup.select_one(".posted_in a")
    category = category_element.get_text(strip=True) if category_element else ""

    # Extract sizes from select options - replace selector as needed
    size_options = soup.select("#size option")
    sizes = [
        opt.get_text(strip=True)
        for opt in size_options
        if opt.get("value") and opt.get("value") != ""
    ]

    # Extract colors from select options - replace selector as needed
    color_options = soup.select("#color option")
    colors = [
        opt.get_text(strip=True)
        for opt in color_options
        if opt.get("value") and opt.get("value") != ""
    ]

    # Extract product images - replace selector as needed
    image_elements = soup.select(".woocommerce-product-gallery__image img")
    images = [img.get("src", "") for img in image_elements if img.get("src")]

    return {
        "title": params["title"],
        "price": params["price"],
        "source_url": params["details_url"],
        "description": description,
        "sku": sku,
        "category": category,
        "sizes": sizes,
        "colors": colors,
        "images": images,
    }


async def automation(
    page: Page,
    params: Params,
    context: BrowserContext | None = None,
    **_kwargs,
):
    """
    Scrapes product details from a product page using BeautifulSoup.
    This API is called via extend_payload from the list API.

    Expected params (from list.py extend_payload):
    {
        "title": "Product Name",
        "price": "$19.99",
        "details_url": "https://www.scrapingcourse.com/ecommerce/product/example"
    }

    This function:
    1. Navigates to the product details page
    2. Extracts additional product information using BeautifulSoup
    3. Returns complete product details
    """
    # Validate required params
    if not params.get("details_url"):
        raise ValueError("details_url is required in params")

    details_url = params["details_url"]
    print(f"Scraping product details from: {details_url}")

    # Navigate to the product page
    await go_to_url(page, details_url)

    # Wait for the product content to load - replace selector as needed
    await page.wait_for_selector(".product")

    # Get the page HTML
    html = await page.content()

    # Extract product details using BeautifulSoup
    product_details = extract_product_details(html, params)

    print(f"Successfully scraped details for: {product_details['title']}")
    return product_details

File: typescript-examples/jsdom-example/api/details.ts
import { BrowserContext, Page } from "playwright";
import { goToUrl } from "@intuned/browser";
import { JSDOM } from "jsdom";
import { z } from "zod";

import { detailsSchema, ProductDetails } from "../utils/typesAndSchemas.js";

type Params = z.infer<typeof detailsSchema>;

function extractProductDetails(html: string, params: Params): ProductDetails {
  /**
   * Extracts detailed product information from the page HTML using JSDOM.
   * Replace selectors with appropriate ones for your target site.
   */
  const dom = new JSDOM(html);
  const document = dom.window.document;

  // Extract description - replace selector as needed
  const descriptionElement = document.querySelector("#tab-description p");
  const description = descriptionElement?.textContent?.trim() || "";

  // Extract SKU - replace selector as needed
  const skuElement = document.querySelector(".sku");
  const sku = skuElement?.textContent?.trim() || "";

  // Extract category - replace selector as needed
  const categoryElement = document.querySelector(".posted_in a");
  const category = categoryElement?.textContent?.trim() || "";

  // Extract sizes from select options - replace selector as needed
  const sizeOptions = document.querySelectorAll("#size option");
  const sizes: string[] = [];
  sizeOptions.forEach((opt) => {
    const value = opt.getAttribute("value");
    if (value && value !== "") {
      sizes.push(opt.textContent?.trim() || "");
    }
  });

  // Extract colors from select options - replace selector as needed
  const colorOptions = document.querySelectorAll("#color option");
  const colors: string[] = [];
  colorOptions.forEach((opt) => {
    const value = opt.getAttribute("value");
    if (value && value !== "") {
      colors.push(opt.textContent?.trim() || "");
    }
  });

  // Extract product images - replace selector as needed
  const imageElements = document.querySelectorAll(
    ".woocommerce-product-gallery__image img"
      );
  const images: string[] = [];
  imageElements.forEach((img) => {
    const src = img.getAttribute("src");
    if (src) {
      images.push(src);
    }
  });

  return {
    title: params.title,
    price: params.price,
    sourceUrl: params.detailsUrl,
    description,
    sku,
    category,
    sizes,
    colors,
    images,
  };
}

export default async function handler(
  params: Params,
  page: Page,
  context: BrowserContext
): Promise<ProductDetails> {
  /**
   * Scrapes product details from a product page using JSDOM.
   * This API is called via extendPayload from the list API.
   *
   * Expected params (from list.ts extendPayload):
   * {
   *   "title": "Product Name",
   *   "price": "$19.99",
   *   "detailsUrl": "https://www.scrapingcourse.com/ecommerce/product/example"
   * }
   *
   * This function:
   * 1. Navigates to the product details page
   * 2. Extracts additional product information using JSDOM
   * 3. Returns complete product details
   */

  // Validate required params
  if (!params.detailsUrl) {
    throw new Error("detailsUrl is required in params");
  }

  const detailsUrl = params.detailsUrl;
  console.log(`Scraping product details from: ${detailsUrl}`);

  // Navigate to the product page
  await goToUrl({ page, url: detailsUrl });

  // Wait for the product content to load - replace selector as needed
  await page.waitForSelector(".product");

  // Get the page HTML
  const html = await page.content();

  // Extract product details using JSDOM
  const productDetails = extractProductDetails(html, params);

  console.log(`Successfully scraped details for: ${productDetails.title}`);
  return productDetails;
}
